compiler/sem/semexprs.nim

```diff
@@ -696,85 +696,135 @@ proc arrayConstrType(c: PContext, n: PNode): PType =
+      let r = fitNode(c, formalIdx, idx, idx.info)
+      if r.kind == nkError:  (r, Zero, nil)
+      else:                  (n[1], getOrdValue(r), formalIdx)
+    elif isOrdinalType(idx.typ):
+      # no formal type is set yet and we got a valid ordinal
+      (n[1], getOrdValue(idx), idx.typ)
+    else:
+      # a constant expression, but not a value usable as an ordinal
+      (c.config.newError(idx, PAstDiag(kind: adSemExpectedOrdinalArrayIdx,
+                                       nonOrdInput: idx,
+                                       indexExpr: n)),
+       Zero, nil)
+  else:
+    (n, Zero, nil)
+
 proc semArrayConstr(c: PContext, n: PNode, flags: TExprFlags): PNode =
   result = newNodeI(nkBracket, n.info)
   result.typ = newTypeS(tyArray, c)
-  rawAddSon(result.typ, nil)     # index type
-  
-  var
-    firstIndex, lastIndex: Int128
-    indexType = getSysType(c.graph, n.info, tyInt)
-    lastValidIndex = lastOrd(c.config, indexType)
-  
+
+  template createRange(first, last: Int128, typ: PType): PType =
+    makeRangeType(c, toInt64(first), toInt64(last), n.info, typ)
+
   if n.len == 0:
+    # an empty array constructor
+    let indexType = getSysType(c.graph, n.info, tyInt)
+    rawAddSon(result.typ, createRange(Zero, toInt128(-1), indexType))
     rawAddSon(result.typ, newTypeS(tyEmpty, c)) # needs an empty basetype!
-    lastIndex = toInt128(-1)
   else:
-    var x = n[0]
-    if x.kind == nkExprColonExpr and x.len == 2:
-      var idx = semConstExpr(c, x[0])
-      if not isOrdinalType(idx.typ):
-        result = n
-        result[0][0] = c.config.newError(x[0],
-                          PAstDiag(kind: adSemExpectedOrdinalArrayIdx,
-                                   nonOrdInput: idx,
-                                   indexExpr: result))
-        # TODO: the error message references the index expression (`result`)
-        #       and we're wrapping the error anyway, maybe expand the nkError
-        #       to cover the whole result in one go
-        result = c.config.wrapError(result)
-        return
-      else:
-        firstIndex = getOrdValue(idx)
-        lastIndex = firstIndex
-        indexType = idx.typ
-        lastValidIndex = lastOrd(c.config, indexType)
-        x = x[1]
-
-    let yy = semExprWithType(c, x)
-    var typ = yy.typ
-    result.add yy
-    for i in 1..<n.len:
-      if lastIndex == lastValidIndex:
-        let validIndex = makeRangeType(
-          c, toInt64(firstIndex),
-          toInt64(lastValidIndex), n.info, indexType)
-
+    # analyse the first element separately -- the other elements depend on its
+    # type
+    var
+      (first, firstIndex, indexType) = semArrayElement(c, n[0], nil)
+      lastIndex = firstIndex ## tracks the index of the previous element
+
+    if indexType == nil:
+      # if no type is provided or an error occurred, default to ``int`` for
+      # the index type
+      indexType = getSysType(c.graph, n.info, tyInt)
```

Comment:
If you'd like to make this a `let` instead of a `var`, you could rename `indexType` to `inferredIndexType`, and then follow that up with a single define/set of `indexType`, like so

```suggestion
    let
      (first, firstIndex, inferredIndexType) = semArrayElement(c, n[0], nil)
      indexType =
        # if no type is provided or an error occurred, default to ``int`` for
        # the index type
        if inferredIndexType.isNil: c.graph.getSysType(n.info, tyInt)
        else: inferredIndexType
                  
    var lastIndex = firstIndex ## tracks the index of the previous element
```

Reply:
It's changed locally, thanks.